### PR TITLE
Materialize a non-default device for C2 legacy storage.

### PR DESCRIPTION
**Pull request by gchanan**
It's not intended that Storages have 'default' CUDA devices, but this is allowable via the Storage::create_legacy codepath.

This also messages with device_caching, because the initial cache is obtained from the Storage, which may have a 'default' device.

Instead, we materialize a device by allocating 0 bytes via the allocator.

### CODE REVIEW

**Comment by gchanan** [comment]
@dzhulgakov does this look correct for C2?

@ezyang can we use this same pattern to address the Sparse issue in https://github.com/pytorch/pytorch/pull/18578?

**Review by dzhulgakov — COMMENTED** [review]
I'm not sure it's worth it. Right now it'd resolve to CUDA allocator and trigger cudaMalloc with 0 bytes. In general, the partially initialized tensor might be created on one device and get its storage only later on another device. So it doesn't solve your problem. I've commented on another PR.

**Comment by ezyang** [comment]
It won't call cudaMalloc. Relevant code:

```
struct CudaCachingAllocator : public Allocator {
  DataPtr allocate(size_t size) const override {
    int device;
    C10_CUDA_CHECK(cudaGetDevice(&device));
    void* r = nullptr;
    if (size != 0) {
      caching_allocator.malloc(&r, size, cuda::getCurrentCUDAStream(device));
    }
    return {r, r, &CudaCachingDeleter, Device(DeviceType::CUDA, device)};
  }
  DeleterFnPtr raw_deleter() const override {
    return &CudaCachingDeleter;
  }
};
```

You'll get the current device but that's it.

**Comment by ezyang** [comment]
> In general, the partially initialized tensor might be created on one device and get its storage only later on another device.

This may very well be true, I don't know enough about Caffe2 to say :)

**Comment by dzhulgakov** [comment]
@ezyang - actually we register this wrapper as the allocator so it might call cudaMalloc under some settings: https://github.com/pytorch/pytorch/blob/master/caffe2/core/context_gpu.cu#L488 (we should probably fix that)

**Comment by gchanan** [comment]
@dzhulgakov I still think

**Review by ezyang — CHANGES_REQUESTED** [review]
Do you want to also fix the unguarded cudaMalloc on zero size site?

File: caffe2/core/context_gpu.cu [hunk 1]
```diff
@@ -485,7 +485,9 @@ struct DefaultCUDAAllocator final : public at::Allocator {
     }
     switch (g_cuda_memory_pool_type) {
       case CudaMemoryPoolType::NONE:
-        CUDA_ENFORCE(cudaMalloc(&ptr, nbytes));
+        if (nbytes != 0) {
+          CUDA_ENFORCE(cudaMalloc(&ptr, nbytes));
+        }
```
Comment (dzhulgakov) [inline caffe2/core/context_gpu.cu]: also CudaMemoryPoolType::CUB: below for completeness?

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**